Improves some text field behavior.

deniz1a opened the pull request:
This patch improves text field behavior in game lobby, settings menu and asset browser.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -272,21 +272,27 @@ public static void SetupEditableNameWidget(Widget parent, Session.Slot s, Sessio
+			name.OnEscKey = () =>
+			{
+				name.Text = c.Name;
+				Ui.Root.Get<TextFieldWidget>("CHAT_TEXTFIELD").TakeKeyboardFocus();
```
This is a generic utility function, so shouldn't make any assumptions about the UI state.  You could instead pass in a `defaultFocus` widget, and then give that focus if it is non-null.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/MapChooserLogic.cs:
```diff
@@ -33,8 +33,18 @@ internal MapChooserLogic(Widget widget, string initialMap, Action onExit, Action
+			{
+				Ui.CloseWindow();
+				onSelect(selectedUid);
+				Ui.Root.Get<TextFieldWidget>("CHAT_TEXTFIELD").TakeKeyboardFocus();
```
The map chooser doesn't have a chat textfield.

deniz1a replied:
This is for when the user closes map chooser. So that the chat text field has focus when returned to lobby window.

deniz1a replied:
Oh ok, I will change map chooser so that it accepts a defaultFocus widget like I did with LobbyUtils.

pchote replied:
Please see my suggestion below, which is a much better implementation strategy than manually adding focus plumbing everywhere.

pchote commented:
I think the idea behind this is good, but the implementation introduces some unsettling coupling.  I think a better approach would be to add a FocusPriority integer to widgets, and then have YieldFocus walk the tree and automatically focus the highest priority (> 0) widget that is visible.

deniz1a commented:
OK I implemented keyboard focus management as a stack. When a widget takes focus, it adds itself on top of the stack. When it yields focus, it removes itself from the stack so the one below it becomes the current focus.

This also allows for automatic focus yielding at widget removal since ForceYieldKeyboardFocus() method is called when a widget is removed. When the map chooser window for example is closed, its focus is removed from the stack and the chat text field of the lobby automatically gets the focus. There is no need to manually yield focus at widget removal or window closure.

Phrohdoh commented on OpenRA.Game/Widgets/Widget.cs:
```diff
@@ -24,9 +24,9 @@ public static class Ui
 		public static int LastTickTime = Game.RunTime;
 
 		static readonly Stack<Widget> WindowList = new Stack<Widget>();
+		static readonly Stack<Widget> KeyboardFocusWidget = new Stack<Widget>();
```
This isn't a widget anymore.

Phrohdoh commented on OpenRA.Game/Widgets/Widget.cs:
```diff
@@ -110,20 +110,54 @@ public static bool HandleInput(MouseInput mi)
 			return Root.HandleTextInputOuter(text);
 		}
 
+		public static bool CheckKeyboardFocus(Widget w)
```
`HasKeyboardFocus` would be a bit better.

deniz1a replied:
OK but the widget class also has a method with the same name. Should I still change it?

deniz1a replied:
No that one is a variable not a method...

Phrohdoh commented on OpenRA.Game/Widgets/Widget.cs:
```diff
@@ -110,20 +110,54 @@ public static bool HandleInput(MouseInput mi)
+			if (KeyboardFocusWidget.Peek() == w)
+				return true;
+
+			return false;
```
If `Peek() == w -> true` is correct, then you can just `return KeyboardFocusWidget.Peek() == w`.

Phrohdoh commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -109,9 +109,14 @@ public AssetBrowserLogic(Widget widget, Action onExit, World world)
+			filenameInput.OnEscKey = () =>
+			{
+				filenameInput.Text = null;
+				filenameInput.OnTextEdited();
```
Are you sure you need to call this?
The text wasn't altered by pressing `esc`.

deniz1a replied:
When Esc is pressed, the text field resets. So the filter has to be updated, otherwise it shows an empty text field but the list is still filtered according to the previous input.

deniz1a commented:
I fixed the issues Phrohdoh pointed out.

deniz1a commented on OpenRA.Game/Widgets/Widget.cs:
```diff
@@ -289,25 +308,34 @@ public virtual bool TakeKeyboardFocus()
 			return true;
 		}
 
 		public virtual bool YieldKeyboardFocus()
```
Why are YieldKeyboardFocus() and TakeKeyboardFocus() virtual? Shouldn't only TryYieldKeyboardFocus() be virtual so a subclass can add conditions for yielding focus?

Mailaender commented:
> This patch improves text field behavior in game lobby, settings menu and asset browser.

Can you elaborate on that? What gets improved/changed?

deniz1a commented:
Here's what this PR does:
* Changes keyboard focus management from a single variable to a stack. When a widget takes keyboard focus, it is added to the stack. When it yields focus, it is removed from stack and the widget below it gets the focus. 
* Asset browser: The filename filter text field gets keyboard focus by default. When the escape key is pressed, text field resets (becomes empty) and the result list gets updated.
* Game lobby: When escape key is pressed, chat text field resets, player name text field resets to unmodified name and keyboard focus is yielded back to chat field.
* Settings menu Display tab: When escape is pressed, frame limit and player name text fields reset to unmodified value and yield focus. When enter is pressed, leading and trailing white spaces are removed from player name input and the value is checked for zero length. If it is zero, text field resets to unmodified value.

deniz1a commented:
Rebased.

deniz1a commented on OpenRA.Game/Widgets/Widget.cs:
```diff
@@ -24,9 +24,9 @@ public static class Ui
 		public static int LastTickTime = Game.RunTime;
 
 		static readonly Stack<Widget> WindowList = new Stack<Widget>();
+		static readonly Stack<Widget> KeyboardFocusList = new Stack<Widget>();
```
Shouldn't these be called WindowStack and KeyboardFocusStack? Since List is another type of collection.

abcdefg30 commented:
Looks good to me. :+1: / :white_check_mark:

pchote commented:
Please let the second review for this be from somebody who can focus on the implementation details.  Our input handling code is one of the most fiddly and important parts of the game, so we need to be careful about changing chunks of it.

Unfortunately I don't have time during the week to spend on reviewing.

deniz1a commented:
I changed the code a bit. Now it also removes null widgets while adding new focus. I added the cleanup method for the following scenario:
Widget is on focus stack but not the current one. Widget gets removed. ForceYieldKeyboardFocus() is called but since it is not on top of the stack and therefore not the current focus, it is not removed from focus list. After removal, widget becomes null? so now we have a null widget on focus list. When the widgets above it yield focus, at a certain time this null widget will get focus which will hold the focus list, no widget below it would ever get focus because it would never yield focus or get removed, since it's null...

deniz1a commented:
I think the current stack implementation is simpler than a linked list.

Mailaender commented on OpenRA.Game/Widgets/Widget.cs:
```diff
@@ -289,25 +322,34 @@ public virtual bool TakeKeyboardFocus()
+			return true;
+		}
+
+		public virtual bool TryYieldKeyboardFocus()
```
This seems to be completely unused outside this class.

deniz1a replied:
Yes that is a template method. A widget can overwrite it and define custom conditions to prevent others from taking focus. This was partially implemented in YieldKeyboardFocus() and used here: https://github.com/OpenRA/OpenRA/pull/8118/files#diff-5a71e6a689956fb5a5f3315e795c9a75L292 . I moved that to this method to make it clearer.

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -190,14 +190,21 @@ Action InitDisplayPanel(Widget panel)
+				frameLimitTextfield.YieldKeyboardFocus();
+				return true;
+			};
+			frameLimitTextfield.OnEscKey = () =>
```
When I press ESC I usually assume the GUI will remove the cursor, not switch to the last selection.

deniz1a replied:
This resets the text field to last saved value and then removes the cursor. Should it save the current value when you press ESC?

penev92 replied:
I'd expect ESC to act like a Cancel button, not like Enter/OK/Apply.

deniz1a replied:
Oh ok now I get what you mean. I'll try to fix that.

deniz1a replied:
@penev92: He means when you click the other text field while one of them has focus and then press ESC the other text field gets focus back.

deniz1a replied:
In fact if you keep clicking on these text fields alternatingly, you add them multiple times to the stack. And then you can keep pressing ESC and the focus switches from one text field to the other.

deniz1a replied:
And in the lobby, if you alternatingly click on chat field and player name field, you are also adding them multiple times to the stack. But there it doesn't cause any problems because chat field doesn't yield focus when ESC is pressed, it just resets the text, so there is no focus switching with ESC. And when the lobby widget is removed, these text fields become null and stay on the stack and get cleaned up with the next YieldKeyboardFocus() call.

chrisforbes replied:
This is pretty weird.

On Mon, Jun 8, 2015 at 10:10 AM, Deniz <notifications@github.com> wrote:

> In OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs
> <https://github.com/OpenRA/OpenRA/pull/8118#discussion_r31881309>:
>
> >  			{
> >  				int fps;
> >  				Exts.TryParseIntegerInvariant(frameLimitTextfield.Text, out fps);
> >  				ds.MaxFramerate = fps.Clamp(1, 1000);
> >  				frameLimitTextfield.Text = ds.MaxFramerate.ToString();
> > +				frameLimitTextfield.YieldKeyboardFocus();
> > +				return true;
> > +			};
> > +			frameLimitTextfield.OnEscKey = () =>
>
> And in the lobby, if you alternatingly click on chat field and player name
> field, you are also adding them multiple times to the stack. But there it
> doesn't cause any problems because chat field doesn't yield focus when ESC
> is pressed, it just resets the text, so there is no focus switching with
> ESC. And when the lobby widget is removed, these text fields become null
> and stay on the stack and get cleaned up with the next YieldKeyboardFocus()
> call.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/8118/files#r31881309>.
>

deniz1a replied:
How should it be done otherwise?

deniz1a replied:
Ok this also happens in Input settings menu with hotkey assignment. If you keep clicking on different hotkey fields, they all get added to the stack and when you press ESC you switch focus in reverse order. So there has to be a way to prevent other text fields in the same widget from getting focus. Or the previous widget should be removed from stack at TakeKeyboardFocus() if they are in the same parent widget?

Mailaender replied:
> How should it be done otherwise?

Convention is that TAB cycles input fields and ESC just leaves them imho.

deniz1a commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -475,16 +496,7 @@ Action InitInputPanel(Widget panel)
-					Ui.KeyboardFocusWidget = null;
-				}
-			};
+			return () => { };
```
Apparently this code block was unnecessary :/

deniz1a commented:
One issue about using ESC key to reset text field is that now it can't be used to close the window. So what should escape key do, reset text field or close window?

penev92 commented:
Depends on the window, really.

deniz1a commented:
Or I could make it like in map chooser window: ESC resets text field if there is text, closes window if text field is empty.

deniz1a committed:
Improves keyboard focus management.
Also improves some text field behavior.

deniz1a commented:
I changed the focus management from stack to widget priority. @pchote was right, it's much better and simpler now.

Mailaender commented:
Not really sure what has changed now. Pressing ESC seems to do the same as now and TAB cycling does nothing. I thought that was the addition here.

deniz1a commented:
Pressing Esc switches text field focus? It should only remove focus and not give it to another text field... And this doesn't implement tab cycling.

Mailaender commented:
I see. ESC seems to do nothing there in current bleed so this works as promised. :white_check_mark:

pchote commented:
This still doesn't look right: you've removed the tracking of the currently focused keyboard widget, and added tracking of the current priority which is a different thing...?  At best this appears to walk the whole widget tree each key press to find the current widget, but at worst there may be cases where this directs input to the wrong widget.

pchote commented:
Reviewers please note that this is a significant change to the core input plumbing, and it must be reviewed based on design as well as function.  "works as promised" isn't sufficient criteria for merging if it hurts performance or generally muddies the engine code.

deniz1a commented:
Yes, for each key press, this walks the whole widget tree but only those that are visible. Currently focused widget is the one that has the highest priority.

I think if the priorities are set correctly and widgets that are not visible have correct IsVisible() value, this wouldn't direct input to wrong widget. Since explicit TakeKeyboardFocus() calls are not necessary now, priority is the only thing that determines focus.

pchote commented:
The idea of having a focus priority assigned for each widget, and using this to decide what widget should be focused when you don't have an explicit assignment is good (and this was what I asked for originally).  Removing the focus tracking, and instead manipulating the widget focus priorities at runtime (and then walking the widget tree for every keyboard event) is not good.

Please reintroduce `TakeKeyboardFocus`, make it so that the focus priorities remain constant with their default or yaml-assigned values, and then use the focus priorities to assign the default focus when opening a new window or yielding an existing focus.

deniz1a commented:
OK, I'll change this like that.

deniz1a commented:
When should ```TakeKeyboardFocus()``` be called?
* Mouse click on text field or hotkey entry widget (Currently this is the only place where it is called)
* When OpenWindow() in Widget.cs is called
* At focus yield

What about tabs? It would also need to be called when active tab is changed. And ingame chat widget would also need to call it I think...

So this would make it more complicated than it currently is. And the benefit would be caching the current focus widget and not checking widget tree at every key press. But is this worth it?

Currently, this code checks the entire ```Ui.Root``` at every key press. But if a widget is not visible, it is skipped and its children are not checked. And ```Ui.Root``` only holds one window at a time. When a new window is opened, the current one is removed from ```Ui.Root``` although it is still in WindowList. So only one window is checked at every key press.

Which one is better, checking highest priority at each key press and eliminating the use of ```TakeKeyboardFocus()``` entirely, or adding ```TakeKeyboardFocus()``` to specific places and checking priority only when needed? The problem with the second option is determining where to add ```TakeKeyboardFocus()```.

deniz1a commented:
And I just realized, in the current version, it is not necessary to call ```ForceYieldKeyboardFocus()``` at widget removal. Because it will already be removed and not appear in widget tree...

deniz1a committed:
Focus yielding is not necessary at widget removal.

deniz1a committed:
Simplification

deniz1a commented:
Actually, focus caching would be implemented with a distinct ```UpdateFocus()``` method and not ```TakeKeyboardFocus()```, because that is an instance method and gives focus to a specific widget. So the problem would be determining when to call ```UpdateFocus()```.

deniz1a commented:
OK, current focus widget must be cached because ```HasKeyboardFocus``` property, which calls ```Ui.GetKeyboardFocus()```, is evaluated every tick by TextFieldWidget here: 
https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Widgets/TextFieldWidget.cs#L251

So... how should I do this for tabs? Manually adding ```Ui.UpdateFocus()``` at every tab.OnClick() seems counter productive.

penev92 commented:
As I understand, this is potentially quite dangerous. Also nobody seems to be willing to review it.
At this point I don't even know what it's meant to do and what problem it solves. I also can't read all the spam above.

Can you please explain in short what this PR does, why, and how, so we can decide if it's something we actually want?

deniz1a commented:
I decided to scrap the focus management idea and opened a different PR with just the text field improvements: #8680.

deniz1a commented:
A priority based automatic focus management would be better but I couldn't do it in a simple way...
